File: src/fable-compiler-js/fable-library/Seq2.js
import { toList, toArray, map, filter, delay } from "./Seq.js";
import { HashSet } from "./MutableSet.js";
import { defaultOf, disposeSafe, getEnumerator } from "./Util.js";
import { addToDict, getItemFromDict, tryGetValue, addToSet } from "./MapUtil.js";
import { Dictionary } from "./MutableMap.js";
import { FSharpRef } from "./Types.js";
export function distinct(xs, comparer) {
    return delay(() => {
        const hashSet = new HashSet([], comparer);
        return filter((x) => addToSet(x, hashSet), xs);
    });
}
export function distinctBy(projection, xs, comparer) {
    return delay(() => {
        const hashSet = new HashSet([], comparer);
        return filter((x) => addToSet(projection(x), hashSet), xs);
    });
}
export function except(itemsToExclude, xs, comparer) {
    return delay(() => {
        const hashSet = new HashSet(itemsToExclude, comparer);
        return filter((x) => addToSet(x, hashSet), xs);
    });
}
export function countBy(projection, xs, comparer) {
    return delay(() => {
        const dict = new Dictionary([], comparer);
        const keys = [];
        const enumerator = getEnumerator(xs);
        try {
            while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                const key = projection(enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]());
                let matchValue;
                let outArg = 0;
                matchValue = [tryGetValue(dict, key, new FSharpRef(() => outArg, (v) => {
                        outArg = (v | 0);
                    })), outArg];
                if (matchValue[0]) {
                    dict.set(key, matchValue[1] + 1);
                }
                else {
                    dict.set(key, 1);
                    void (keys.push(key));
                }
            }
        }
        finally {
            disposeSafe(enumerator);
        }
        return map((key_1) => [key_1, getItemFromDict(dict, key_1)], keys);
    });
}
export function groupBy(projection, xs, comparer) {
    return delay(() => {
        const dict = new Dictionary([], comparer);
        const keys = [];
        const enumerator = getEnumerator(xs);
        try {
            while (enumerator["System.Collections.IEnumerator.MoveNext"]()) {
                const x = enumerator["System.Collections.Generic.IEnumerator`1.get_Current"]();
                const key = projection(x);
                let matchValue;
                let outArg = defaultOf();
                matchValue = [tryGetValue(dict, key, new FSharpRef(() => outArg, (v) => {
                        outArg = v;
                    })), outArg];
                if (matchValue[0]) {
                    void (matchValue[1].push(x));
                }
                else {
                    addToDict(dict, key, [x]);
                    void (keys.push(key));
                }
            }
        }
        finally {
            disposeSafe(enumerator);
        }
        return map((key_1) => [key_1, getItemFromDict(dict, key_1)], keys);
    });
}
export function Array_distinct(xs, comparer) {
    return toArray(distinct(xs, comparer));
}
export function Array_distinctBy(projection, xs, comparer) {
    return toArray(distinctBy(projection, xs, comparer));
}
export function Array_except(itemsToExclude, xs, comparer) {
    return toArray(except(itemsToExclude, xs, comparer));
}
export function Array_countBy(projection, xs, comparer) {
    return toArray(countBy(projection, xs, comparer));
}
export function Array_groupBy(projection, xs, comparer) {
    return toArray(map((tupledArg) => [tupledArg[0], toArray(tupledArg[1])], groupBy(projection, xs, comparer)));
}
export function List_distinct(xs, comparer) {
    return toList(distinct(xs, comparer));
}
export function List_distinctBy(projection, xs, comparer) {
    return toList(distinctBy(projection, xs, comparer));
}
export function List_except(itemsToExclude, xs, comparer) {
    return toList(except(itemsToExclude, xs, comparer));
}
export function List_countBy(projection, xs, comparer) {
    return toList(countBy(projection, xs, comparer));
}
export function List_groupBy(projection, xs, comparer) {
    return toList(map((tupledArg) => [tupledArg[0], toList(tupledArg[1])], groupBy(projection, xs, comparer)));
}

File: src/fable-library-dart/Seq2.fs
// split from Seq to remove circular dependencies
module SeqModule2

open Fable.Core

let distinct
    (xs: seq<'T>)
    ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'T>)
    =
    Seq.delay (fun () ->
        let hashSet = System.Collections.Generic.HashSet<'T>(comparer)
        xs |> Seq.filter (fun x -> hashSet.Add(x))
    )

let distinctBy
    (projection: 'T -> 'Key)
    (xs: seq<'T>)
    ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
    =
    Seq.delay (fun () ->
        let hashSet = System.Collections.Generic.HashSet<'Key>(comparer)
        xs |> Seq.filter (fun x -> hashSet.Add(projection x))
    )

let except
    (itemsToExclude: seq<'T>)
    (xs: seq<'T>)
    ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'T>)
    =
    Seq.delay (fun () ->
        let hashSet =
            System.Collections.Generic.HashSet<'T>(itemsToExclude, comparer)

        xs |> Seq.filter (fun x -> hashSet.Add(x))
    )

let countBy
    (projection: 'T -> 'Key)
    (xs: seq<'T>)
    ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
    : ('Key * int) seq
    =
    Seq.delay (fun () ->
        let dict = System.Collections.Generic.Dictionary<'Key, int>(comparer)
        let keys = ResizeArray<'Key>()

        for x in xs do
            let key = projection x

            match dict.TryGetValue key with
            | true, v -> dict[key] <- v + 1
            | false, _ ->
                dict[key] <- 1
                keys.Add(key)

        Seq.map (fun key -> key, dict[key]) keys
    )

let groupBy
    (projection: 'T -> 'Key)
    (xs: seq<'T>)
    ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
    : ('Key * seq<'T>) seq
    =
    Seq.delay (fun () ->
        let dict =
            System.Collections.Generic.Dictionary<'Key, ResizeArray<'T>>(
                comparer
            )

        let keys = ResizeArray<'Key>()

        for x in xs do
            let key = projection x

            if dict.ContainsKey(key) then
                dict[key].Add(x)
            else
                dict[key] <- ResizeArray [ x ]
                keys.Add(key)

        Seq.map (fun key -> key, dict[key] :> seq<'T>) keys
    )

module Array =

    let distinct
        (xs: 'T[])
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'T>)
        : 'T[]
        =
        distinct xs comparer |> Seq.toArray

    let distinctBy
        (projection: 'T -> 'Key)
        (xs: 'T[])
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
        : 'T[]
        =
        distinctBy projection xs comparer |> Seq.toArray

    let except
        (itemsToExclude: seq<'T>)
        (xs: 'T[])
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'T>)
        : 'T[]
        =
        except itemsToExclude xs comparer |> Seq.toArray

    let countBy
        (projection: 'T -> 'Key)
        (xs: 'T[])
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
        : ('Key * int)[]
        =
        countBy projection xs comparer |> Seq.toArray

    let groupBy
        (projection: 'T -> 'Key)
        (xs: 'T[])
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
        : ('Key * 'T[])[]
        =
        groupBy projection xs comparer
        |> Seq.map (fun (key, values) -> key, Seq.toArray values)
        |> Seq.toArray

module List =

    let distinct
        (xs: 'T list)
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'T>)
        : 'T list
        =
        distinct xs comparer |> Seq.toList

    let distinctBy
        (projection: 'T -> 'Key)
        (xs: 'T list)
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
        : 'T list
        =
        distinctBy projection xs comparer |> Seq.toList

    let except
        (itemsToExclude: seq<'T>)
        (xs: 'T list)
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'T>)
        : 'T list
        =
        except itemsToExclude xs comparer |> Seq.toList

    let countBy
        (projection: 'T -> 'Key)
        (xs: 'T list)
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
        : ('Key * int) list
        =
        countBy projection xs comparer |> Seq.toList

    let groupBy
        (projection: 'T -> 'Key)
        (xs: 'T list)
        ([<Inject>] comparer: System.Collections.Generic.IEqualityComparer<'Key>)
        : ('Key * 'T list) list
        =
        groupBy projection xs comparer
        |> Seq.map (fun (key, values) -> key, Seq.toList values)
        |> Seq.toList
